core/src/main/kotlin/com/tschuchort/compiletesting/KotlinJsCompilation.kt

```diff
@@ -0,0 +1,139 @@
+  var kotlinStdLibJsJar: File? by default {
+    findInHostClasspath(hostClasspaths, "kotlin-stdlib-js.jar",
+      kotlinDependencyRegex("kotlin-stdlib-js"))
+  }
+
+  // *.class files, Jars and resources (non-temporary) that are created by the
+  // compilation will land here
+  val outputDir get() = workingDir.resolve("output")
+
+  /** Result of the compilation */
+  inner class Result(
+    /** The exit code of the compilation */
+    val exitCode: KotlinCompilation.ExitCode,
+    /** Messages that were printed by the compilation */
+    val messages: String
+  ) {
+    /** The directory where only the final output class and resources files will be */
+    val outputDirectory: File get() = outputDir
+
+    /**
+     * Compiled class and resource files that are the final result of the compilation.
+     */
+    val compiledClassAndResourceFiles: List<File> = outputDirectory.listFilesRecursively()
+  }
+
+
+  // setup common arguments for the two kotlinc calls
+  private fun commonK2JSArgs() = commonArguments(K2JSCompilerArguments()) { args ->
+    // the compiler should never look for stdlib or reflect in the
+    // kotlinHome directory (which is null anyway). We will put them
+    // in the classpath manually if they're needed
+    args.noStdlib = true
+
+    args.moduleKind = "commonjs"
+    args.outputFile = File(outputDir, outputFileName).absolutePath
+    args.sourceMapBaseDirs = commonClasspaths().joinToString(separator = File.pathSeparator)
+    args.libraries = listOfNotNull(kotlinStdLibJsJar).joinToString(separator = ":")
+
+    args.irProduceKlibDir = irProduceKlibDir
+    args.irProduceKlibFile = irProduceKlibFile
+    args.irProduceJs = irProduceJs
+    args.irDce = irDce
+    args.irDceDriven = irDceDriven
+    args.irDcePrintReachabilityInfo = irDcePrintReachabilityInfo
+    args.irOnly = irOnly
+    args.irModuleName = irModuleName
+  }
+
+  /** Runs the compilation task */
+  fun compile(): Result {
+    // make sure all needed directories exist
+    sourcesDir.mkdirs()
+    outputDir.mkdirs()
+
+    // write given sources to working directory
+    val sourceFiles = sources.map { it.writeIfNeeded(sourcesDir) }
+
+    pluginClasspaths.forEach { filepath ->
+      if (!filepath.exists()) {
+        error("Plugin $filepath not found")
+        return makeResult(KotlinCompilation.ExitCode.INTERNAL_ERROR)
+      }
+    }
+
+
+    /* Work around for warning that sometimes happens:
+    "Failed to initialize native filesystem for Windows
+    java.lang.RuntimeException: Could not find installation home path.
+    Please make sure bin/idea.properties is present in the installation directory"
+    See: https://github.com/arturbosch/detekt/issues/630
+    */
+    withSystemProperty("idea.use.native.fs.for.win", "false") {
+      // step 1: compile Kotlin files
+      return makeResult(compileKotlin(sourceFiles, K2JSCompiler(), commonK2JSArgs()))
+    }
+  }
+
+  private fun makeResult(exitCode: KotlinCompilation.ExitCode): Result {
```

Comment:
Originally I pulled ExitCode out as a separate class but then realized that breaks compatibility so put it back. I wanted to share this Enum between compilations since there really isn't a reason for separate ones. But I understand it is a little awkward. Should I duplicate it?